lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -171,6 +273,198 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
                          M.getASTContext());
 }
 
+std::pair<GenericEnvironment *, GenericSignature *>
+getSignatureWithRequirements(GenericSignature *OrigGenSig,
+                             GenericEnvironment *OrigGenericEnv,
+                             ArrayRef<RequirementRepr> Requirements,
+                             SILModule &M) {
+  // Form a new generic signature based on the old one.
+  ArchetypeBuilder Builder(M.getASTContext(),
+                           LookUpConformanceInModule(M.getSwiftModule()));
+
+  // First, add the old generic signature.
+  Builder.addGenericSignature(OrigGenSig);
+
+  // For each substitution with a concrete type as a replacement,
+  // add a new concrete type equality requirement.
+  for (auto &Req : Requirements) {
+    Builder.addRequirement(Req);
+  }
+
+  auto *GenericSig = Builder.getGenericSignature();
+  // Remember the new generic environment.
+  auto *GenericEnv = Builder.getGenericEnvironment(GenericSig);
+
+  return std::make_pair(GenericEnv, GenericSig);
+}
+
+/// Perform some sanity checks for the requirements
+static void
+checkSpecializationRequirements(ArrayRef<RequirementRepr> Requirements) {
+  for (auto &Req : Requirements) {
+    if (Req.getKind() == RequirementReprKind::SameType) {
+      auto FirstType = Req.getFirstType();
+      auto SecondType = Req.getSecondType();
+      assert(FirstType && SecondType);
+
+      bool isFirstTypeNonConcrete =
+          FirstType->hasArchetype() || FirstType->hasTypeParameter();
+      bool isSecondTypeNonConcrete =
+          SecondType->hasArchetype() || SecondType->hasTypeParameter();
+      // Only one of the types should be concrete.
+      assert((isFirstTypeNonConcrete ^ isSecondTypeNonConcrete) &&
+             "Only concrete type same-type requirements are supported by "
+             "generic specialization");
+      continue;
+    }
+
+    if (Req.getKind() == RequirementReprKind::LayoutConstraint) {
+      // TODO: Check that it is one of the compiler known protocols used for
+      // pre-specializations.
+      continue;
+    }
+
+    // TODO: Extend it to support TypeConstraint requirements once it is
+    // supported?
+    llvm_unreachable("Unknown type of requirement in generic specialization");
+  }
+}
+
+/// Remap Requirements to RequirementReps.
```

Comment:
Good catch. I've written this as a workaround a while ago and it is most likely outdated. IIRC, there were some issues with the ArchetypeBuilder APIs at that time. Somehow, only RequirementReprs could be added to the builder using the APIs or something like this. But we don't need this hack anymore. I'll remove it if everything builds OK without it.